### ISSUE TITLE
Accuracy gap between tester and InferenceModal

### ISSUE THREAD

**Issue by Penn000** [KevinMusgrave/pytorch-metric-learning#356]
Hi, I'm doing a classification task via metric learning, and follow the [TrainWithClassifier.ipynb](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/examples/notebooks/TrainWithClassifier.ipynb) to train my own dataset. tester report a relatively high accuracy in log.

```python
def train(**kwargs):    
    # model
    # model as trunk
    Model = model_factory[cfg.model_name]
    trunk = Model(cfg)
    trunk = torch.nn.DataParallel(trunk.to(device))
    # Set embedder model. 
    embedder = torch.nn.DataParallel(MLP([cfg.feat_size, cfg.embed_size]).to(device))
    # Set the classifier
    classifier = torch.nn.DataParallel(MLP([cfg.embed_size, cfg.n_classes])).to(device)
    
    # Set the loss function
    loss = losses.TripletMarginLoss(margin=cfg.margin)
    # Set the classification loss:
    classification_loss = torch.nn.CrossEntropyLoss()
    
    # Set the dataset
    train_dataset = TrainDataset(cfg)
    val_dataset = TestDataset(cfg)
    # Set the mining function
    miner = miners.MultiSimilarityMiner(epsilon=0.1)
    # Set the dataloader sampler
    sampler = samplers.MPerClassSampler(train_dataset.labellist, m=4, length_before_new_iter=len(train_dataset))
    
    # Set optimizers
    trunk_optimizer      = torch.optim.Adam(trunk.parameters(),      lr=cfg.lr_start, weight_decay=cfg.weight_decay)
    embedder_optimizer   = torch.optim.Adam(embedder.parameters(),   lr=cfg.lr_start, weight_decay=cfg.weight_decay)
    classifier_optimizer = torch.optim.Adam(classifier.parameters(), lr=cfg.lr_start, weight_decay=cfg.weight_decay)


    # Package the above stuff into dictionaries.
    models = {"trunk": trunk, "embedder": embedder, "classifier": classifier}
    optimizers = {"trunk_optimizer": trunk_optimizer, "embedder_optimizer": embedder_optimizer, "classifier_optimizer": classifier_optimizer}
    loss_funcs = {"metric_loss": loss, "classifier_loss": classification_loss}
    mining_funcs = {"tuple_miner": miner}

    # We can specify loss weights if we want to. This is optional
    loss_weights = {"metric_loss": 1, "classifier_loss": 0.5}
    
    record_keeper, _, _ = logging_presets.get_record_keeper(cfg.respth, cfg.respth)
    hooks = logging_presets.get_hook_container(record_keeper)
    dataset_dict = {"val": val_dataset}
    tester = testers.GlobalEmbeddingSpaceTester(end_of_testing_hook = hooks.end_of_testing_hook, 
                    dataloader_num_workers = 4,
                    accuracy_calculator=AccuracyCalculator(k="max_bin_count"))
    end_of_epoch_hook = hooks.end_of_epoch_hook(tester, 
                                                dataset_dict, 
                                                cfg.respth, 
                                                test_interval = 1,
                                                patience = 1)
    
    trainer = trainers.TrainWithClassifier(models,
                      optimizers,
                      cfg.batch_size,
                      loss_funcs,
                      mining_funcs,
                      train_dataset,
                      sampler=sampler,
                      dataloader_num_workers = 4,
                      loss_weights = loss_weights,
                      end_of_iteration_hook = hooks.end_of_iteration_hook,
                      end_of_epoch_hook = end_of_epoch_hook)
    
    trainer.train(num_epochs=cfg.num_epochs)
```

but when I evaluate the same test dataset via InferenceModal, the accuracy is very low

```python
def test():
    # dataset
    train_ds = TrainDataset(cfg)
    test_ds = TestDataset(cfg)
    test_dl = DataLoader(test_ds, batch_size=1)
    
    # model
    # logo model as trunk
    LogoModel = model_factory[cfg.model_name]
    trunk = LogoModel(cfg).to(device)
    trunk.load_state_dict(torch.load('exp/trunk_best1.pth'))
    # Set embedder model. 
    embedder = MLP([cfg.feat_size, cfg.embed_size]).to(device)
    embedder.load_state_dict(torch.load('exp/embedder_best1.pth'))
    
    match_finder = MatchFinder(distance=LpDistance(normalize_embeddings=False, p=2, power=1))
    inference_model = InferenceModel(trunk, embedder, match_finder=match_finder)
    inference_model.train_indexer(train_ds, batch_size=256)
    print('train indexer done.')
        
    tp, fp, tn, fn = 0, 0, 0, 0
    for imgs, label in test_dl:
        imgs = imgs.to(device)
        
        idxs, dis = inference_model.get_nearest_neighbors(imgs, k=1)
        idxs = idxs.reshape(-1)
        pred = train_ds.labellist[idxs[0]]
        
        if pred != -1: # positive prediction
            if pred == label: tp += 1
            else: fp += 1
        else: # negative prediction
            if len(label) == 0: tn += 1
            else: fn += 1
        
    print('TP: {}, FP: {}, TN: {}, FN: {}'.format(tp, fp, tn, fn))
    print('Precision: {}, Recall: {}'.format(tp/(tp+fp), tp/(tp+fn)))
```

Is there any details I ignored ? Could you tell me the reason why accuracy gap between tester and InferenceModal is large? 
Thank you in advance.

**Comment by KevinMusgrave** [comment]
My first guess was that your MatchFinder is using ```normalize_embeddings=False```, while everything is using ```normalize_embeddings=True```. But I realized that MatchFinder is only used if you call ```inference_model.get_matches``` or ```inference_model.is_match```. Confusing, I know 😕. Anyway I would set that to True just to avoid confusion if you do end up calling those methods. 
```python
# set normalize_embeddings to True 
match_finder = MatchFinder(distance=LpDistance(normalize_embeddings=True, p=2, power=1))
```

Re: accuracy gap, there could be a couple of reasons.

1) [By default](https://kevinmusgrave.github.io/pytorch-metric-learning/testers/#testing-splits), the tester is going to use each dataset in ```dataset_dict``` as its own reference set. If you want to use your training set as the reference, then you should do:

```python
splits_to_eval = [('val', ['train'])]
end_of_epoch_hook = hooks.end_of_epoch_hook(tester, 
                                                dataset_dict, 
                                                cfg.respth, 
                                                test_interval = 1,
                                                patience = 1,
                                                splits_to_eval = splits_to_eval)
```

2) The way you compute precision:

```python
        if pred != -1: # positive prediction
            if pred == label: tp += 1
            else: fp += 1
        else: # negative prediction
            if len(label) == 0: tn += 1
            else: fn += 1
```

To get the same ```precision_at_1``` as AccuracyCalculator, I think you should do:

```python
if pred == label: tp += 1

# after the for loop
precision_at_1 = tp / len(test_ds)
```

**Comment by Penn000** [comment]
Thanks for your reply.  I set `normalize_embeddings=True`  in MatchFinder at first and it did not work as you explain the reason. By the way, I wonder which distance is used in `inference_model.get_nearest_neighbors`? This function seems could not pass in distance as a parameter.


> 1. By default, the tester is going to use each dataset in dataset_dict as its own reference set. If you want to use your training set as the reference, then you should do...

I think this is the main reason for accuracy gap, because when I check the log I find `INFO base_tester.py(299): Computing accuracy for the val split w.r.t ['val']`. I retrain the model follow your advice, and now the accuracy of tester and InferenceModel is both low :joy: .

```python
dataset_dict = {"val": val_dataset, 'train': train_dataset}
splits_to_eval = [('val', ['train'])]
end_of_epoch_hook = hooks.end_of_epoch_hook(tester, 
                                                dataset_dict, 
                                                cfg.respth, 
                                                test_interval = 1,
                                                patience = 1,
                                                splits_to_eval = splits_to_eval)
```

> 2. The way you compute precision...

I guess `precision_at_1 ` in AccuracyCalculator is equal to `(tp+tn)/(tp+fp+tn+fn)`. My trouble is that the number of tp is too little.

**Comment by KevinMusgrave** [comment]
> Thanks for your reply. I set `normalize_embeddings=True` in MatchFinder at first and it did not work as you explain the reason. By the way, I wonder which distance is used in `inference_model.get_nearest_neighbors`? This function seems could not pass in distance as a parameter.

It uses the L2 distance in [FaissIndexer](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/utils/inference.py#L51-L73). InferenceModel also takes in a ```normalize_embeddings``` argument, so basically you can do normalized L2 or unnormalized L2. Or you can create your own FaissIndexer and use some other index (see the [faiss repo](https://github.com/facebookresearch/faiss)), and pass it into InferenceModel as ```indexer```.

> I guess `precision_at_1 ` in AccuracyCalculator is equal to `(tp+tn)/(tp+fp+tn+fn)`. My trouble is that the number of tp is too little.

If there is a big class imbalance, you can set ```avg_of_avgs=True``` in AccuracyCalculator. This will compute the average accuracy per class, and return the average of those averages.

```python
AccuracyCalculator(k="max_bin_count", avg_of_avgs=True)
```

**Comment by Penn000** [comment]
Thanks a lot! your suggestion is very helpful
